In [ ]:
import torch
print(torch.__version__)

2.6.0+cu124


In [ ]:
if torch.cuda.is_available():
    print("GPU is available!")
    print(f"Using GPU: {torch.cuda.get_device_name(0)}")
else:
    print("GPU not available. Using CPU.")

GPU is available!
Using GPU: Tesla T4


## Creating a Tensor

In [ ]:
# using empty
a = torch.empty(2,3)

### Explanation of `torch.empty(2, 3)`

- `torch.empty(2, 3)` creates a **2x3 tensor** (2 rows and 3 columns) without initializing its values.
- The values are **uninitialized** and contain whatever data was already in that allocated memory, so it appears as random garbage values.
- This is faster than filling the tensor with zeros or ones since it skips explicit initialization.


In [ ]:
# check type
type(a)

torch.Tensor

### Explanation of `type(a)`

- `type(a)` is used to **check the data type** of the variable `a`.
- Since `a` is a tensor created with `torch.empty(2, 3)`, its type will be:
  - `<class 'torch.Tensor'>`
- This confirms that the object is indeed a **PyTorch Tensor**.


In [ ]:
# using zeros
torch.zeros(2,3)

tensor([[0., 0., 0.],
        [0., 0., 0.]])

### Explanation of `torch.zeros(2, 3)`

- `torch.zeros(2, 3)` creates a **2x3 tensor** (2 rows and 3 columns) where **all elements are initialized to 0**.
- This is useful when you need a tensor of a specific size with a known initial state.
- The data type of the tensor defaults to `float32`, but it can be changed using the `dtype` parameter if needed.


In [ ]:
# using ones
torch.ones(2,3)

tensor([[1., 1., 1.],
        [1., 1., 1.]])

### Explanation of `torch.ones(2, 3)`

- `torch.ones(2, 3)` creates a **2x3 tensor** (2 rows and 3 columns) where **all elements are initialized to 1**.
- This is useful when you need a tensor of a specific size with a known initial value.
- The data type of the tensor defaults to `float32`, but it can be changed using the `dtype` parameter if needed.


In [ ]:
# using rand
torch.rand(2,3)

tensor([[0.1127, 0.5653, 0.1766],
        [0.3267, 0.5564, 0.9091]])

### Explanation of `torch.rand(2, 3)`

- `torch.rand(2, 3)` creates a **2x3 tensor** (2 rows and 3 columns) where **all elements are initialized with random values** drawn from a uniform distribution in the range [0, 1).
- This is useful when you need a tensor with random values for operations like weight initialization in neural networks.
- The data type of the tensor defaults to `float32`, but it can be changed using the `dtype` parameter if needed.


In [ ]:
# use of seed
torch.rand(2,3)

tensor([[0.3105, 0.5873, 0.9429],
        [0.5641, 0.2928, 0.9380]])

### Explanation of `torch.rand(2, 3)`

- `torch.rand(2, 3)` creates a **2x3 tensor** (2 rows and 3 columns) where **all elements are initialized with random values** drawn from a uniform distribution in the range [0, 1).
- This is useful when you need a tensor with random values, often used for initializing parameters like weights in neural networks.
- The data type of the tensor defaults to `float32`, but it can be changed using the `dtype` parameter if needed.


In [ ]:
# manual_seed
torch.manual_seed(100)
torch.rand(2,3)

tensor([[0.1117, 0.8158, 0.2626],
        [0.4839, 0.6765, 0.7539]])

### Explanation of `torch.manual_seed(100)` and `torch.rand(2, 3)`

- `torch.manual_seed(100)` sets the random seed for generating random numbers, ensuring reproducibility of the random values generated in subsequent operations.
- `torch.rand(2, 3)` creates a **2x3 tensor** (2 rows and 3 columns) where **all elements are initialized with random values** drawn from a uniform distribution in the range [0, 1).
- By setting the random seed with `torch.manual_seed(100)`, the random values generated by `torch.rand(2, 3)` will be the same every time the code is run, ensuring consistent results.
- The data type of the tensor defaults to `float32`, but it can be changed using the `dtype` parameter if needed.


In [ ]:
torch.manual_seed(100)
torch.rand(2,3)

tensor([[0.1117, 0.8158, 0.2626],
        [0.4839, 0.6765, 0.7539]])

In [ ]:
# using tensor
torch.tensor([[1,2,3],[4,5,6]])

tensor([[1, 2, 3],
        [4, 5, 6]])

### Explanation of `torch.tensor([[1, 2, 3], [4, 5, 6]])`

- `torch.tensor([[1, 2, 3], [4, 5, 6]])` creates a **2x3 tensor** (2 rows and 3 columns) with the specified values.
- The tensor is directly initialized with the values provided in the list of lists `[[1, 2, 3], [4, 5, 6]]`.
- The data type of the tensor defaults to `int64` based on the input values, but it can be changed using the `dtype` parameter if needed.


In [ ]:
# other ways

# arange
print("using arange ->", torch.arange(0,10,2))

# using linspace
print("using linspace ->", torch.linspace(0,10,10))

# using eye
print("using eye ->", torch.eye(5))

# using full
print("using full ->", torch.full((3, 3), 5))

using arange -> tensor([0, 2, 4, 6, 8])
using linspace -> tensor([ 0.0000,  1.1111,  2.2222,  3.3333,  4.4444,  5.5556,  6.6667,  7.7778,
         8.8889, 10.0000])
using eye -> tensor([[1., 0., 0., 0., 0.],
        [0., 1., 0., 0., 0.],
        [0., 0., 1., 0., 0.],
        [0., 0., 0., 1., 0.],
        [0., 0., 0., 0., 1.]])
using full -> tensor([[5, 5, 5],
        [5, 5, 5],
        [5, 5, 5]])


### Explanation of `torch.arange`, `torch.linspace`, `torch.eye`, and `torch.full`

- **Using `torch.arange(0, 10, 2)`**:
  - Creates a **1D tensor** with values starting from 0 up to, but not including, 10, with a step size of 2.
  - The output tensor is `[0, 2, 4, 6, 8]`.

- **Using `torch.linspace(0, 10, 10)`**:
  - Creates a **1D tensor** with 10 equally spaced values between 0 and 10, inclusive.
  - The output tensor is `[0.0000, 1.1111, 2.2222, 3.3333, 4.4444, 5.5556, 6.6667, 7.7778, 8.8889, 10.0000]`.

- **Using `torch.eye(5)`**:
  - Creates a **5x5 identity matrix** where the diagonal elements are 1, and all other elements are 0.
  - The output tensor is:
    ```
    [[1., 0., 0., 0., 0.],
     [0., 1., 0., 0., 0.],
     [0., 0., 1., 0., 0.],
     [0., 0., 0., 1., 0.],
     [0., 0., 0., 0., 1.]]
    ```

- **Using `torch.full((3, 3), 5)`**:
  - Creates a **3x3 tensor** where all elements are initialized with the value 5.
  - The output tensor is:
    ```
    [[5, 5, 5],
     [5, 5, 5],
     [5, 5, 5]]
    ```


## Tensor Shapes

In [ ]:
x = torch.tensor([[1,2,3],[4,5,6]])
x

tensor([[1, 2, 3],
        [4, 5, 6]])

### Explanation of `torch.tensor([[1, 2, 3], [4, 5, 6]])`

- `torch.tensor([[1, 2, 3], [4, 5, 6]])` creates a **2x3 tensor** (2 rows and 3 columns) with the specified values.
- The tensor is directly initialized with the values provided in the list of lists `[[1, 2, 3], [4, 5, 6]]`.
- The data type of the tensor defaults to `int64` based on the input values, but it can be changed using the `dtype` parameter if needed.

Output:


In [ ]:
x.shape

torch.Size([2, 3])

In [ ]:
torch.empty_like(x)

tensor([[                  0, 7235419174270214779, 3761406417083316770],
        [3544385878256673633, 7292279102833320243, 7161347252428484965]])

### Explanation of `torch.empty_like(x)`

- `torch.empty_like(x)` creates a **new tensor** with the **same shape and type as `x`**, but the values are **uninitialized**.
- The contents of the tensor are whatever values were already present at that memory location, which is why you may see random numbers.
- This is useful for creating a tensor of the same structure without predefined values, typically for in-place operations.



In [ ]:
torch.zeros_like(x)

tensor([[0, 0, 0],
        [0, 0, 0]])

In [ ]:
torch.ones_like(x)

tensor([[1, 1, 1],
        [1, 1, 1]])

### Explanation of `torch.ones_like(x)`

- `torch.ones_like(x)` creates a **new tensor** with the **same shape and type as `x`**, but **all elements are initialized to 1**.
- The structure (number of rows and columns) matches that of `x`, and the data type is inherited unless explicitly specified with `dtype`.


In [ ]:
torch.rand_like(x, dtype=torch.float32)

tensor([[0.2627, 0.0428, 0.2080],
        [0.1180, 0.1217, 0.7356]])

### Explanation of `torch.rand_like(x, dtype=torch.float32)`

- `torch.rand_like(x, dtype=torch.float32)` creates a **new tensor** with the **same shape as `x`** but with **random values** drawn from a **uniform distribution** in the range [0, 1).
- The `dtype` parameter is explicitly set to `torch.float32`, ensuring the tensor values are in 32-bit floating point.


## Tensor Data Types

In [ ]:
# find data type
x.dtype

torch.int64

In [ ]:
# assign data type
torch.tensor([1.0,2.0,3.0], dtype=torch.int32)

tensor([1, 2, 3], dtype=torch.int32)

### Explanation of `torch.tensor([1.0, 2.0, 3.0], dtype=torch.int32)`

- `torch.tensor([1.0, 2.0, 3.0], dtype=torch.int32)` creates a **1D tensor** with the specified values `[1.0, 2.0, 3.0]`.
- The `dtype` parameter is explicitly set to `torch.int32`, which means the floating-point values are **cast to integers** during tensor creation.
- As a result, the tensor will store the values as integers, discarding the decimal points.



In [ ]:
torch.tensor([1,2,3], dtype=torch.float64)

tensor([1., 2., 3.], dtype=torch.float64)

In [ ]:
# using to()
x.to(torch.float32)

tensor([[1., 2., 3.],
        [4., 5., 6.]])

### Explanation of `x.to(torch.float32)`

- `x.to(torch.float32)` converts the tensor `x` to the **data type `float32`**.
- This operation does not modify the original tensor but instead **returns a new tensor** with the specified data type.
- It is commonly used when you need to change the data type of a tensor for computations that require specific precision.



| **Data Type**             | **Dtype**         | **Description**                                                                                                                                                                |
|---------------------------|-------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **32-bit Floating Point** | `torch.float32`   | Standard floating-point type used for most deep learning tasks. Provides a balance between precision and memory usage.                                                         |
| **64-bit Floating Point** | `torch.float64`   | Double-precision floating point. Useful for high-precision numerical tasks but uses more memory.                                                                               |
| **16-bit Floating Point** | `torch.float16`   | Half-precision floating point. Commonly used in mixed-precision training to reduce memory and computational overhead on modern GPUs.                                            |
| **BFloat16**              | `torch.bfloat16`  | Brain floating-point format with reduced precision compared to `float16`. Used in mixed-precision training, especially on TPUs.                                                |
| **8-bit Floating Point**  | `torch.float8`    | Ultra-low-precision floating point. Used for experimental applications and extreme memory-constrained environments (less common).                                               |
| **8-bit Integer**         | `torch.int8`      | 8-bit signed integer. Used for quantized models to save memory and computation in inference.                                                                                   |
| **16-bit Integer**        | `torch.int16`     | 16-bit signed integer. Useful for special numerical tasks requiring intermediate precision.                                                                                    |
| **32-bit Integer**        | `torch.int32`     | Standard signed integer type. Commonly used for indexing and general-purpose numerical tasks.                                                                                  |
| **64-bit Integer**        | `torch.int64`     | Long integer type. Often used for large indexing arrays or for tasks involving large numbers.                                                                                  |
| **8-bit Unsigned Integer**| `torch.uint8`     | 8-bit unsigned integer. Commonly used for image data (e.g., pixel values between 0 and 255).                                                                                    |
| **Boolean**               | `torch.bool`      | Boolean type, stores `True` or `False` values. Often used for masks in logical operations.                                                                                      |
| **Complex 64**            | `torch.complex64` | Complex number type with 32-bit real and 32-bit imaginary parts. Used for scientific and signal processing tasks.                                                               |
| **Complex 128**           | `torch.complex128`| Complex number type with 64-bit real and 64-bit imaginary parts. Offers higher precision but uses more memory.                                                                 |
| **Quantized Integer**     | `torch.qint8`     | Quantized signed 8-bit integer. Used in quantized models for efficient inference.                                                                                              |
| **Quantized Unsigned Integer** | `torch.quint8` | Quantized unsigned 8-bit integer. Often used for quantized tensors in image-related tasks.                                                                                     |


## Mathematical operations

### 1. Scalar operation

In [ ]:
x = torch.rand(2,2)
x

tensor([[0.7118, 0.7876],
        [0.4183, 0.9014]])

In [ ]:
# addition
x + 2
# substraction
x - 2
# multiplication
x * 3
# division
x / 3
# int division
(x * 100)//3
# mod
((x * 100)//3)%2
# power
x**2

tensor([[0.5066, 0.6203],
        [0.1750, 0.8125]])

### 2. Element wise operation

In [ ]:
a = torch.rand(2,3)
b = torch.rand(2,3)

print(a)
print(b)

tensor([[0.9969, 0.7565, 0.2239],
        [0.3023, 0.1784, 0.8238]])
tensor([[0.5557, 0.9770, 0.4440],
        [0.9478, 0.7445, 0.4892]])


In [ ]:
# add
a + b
# sub
a - b
# multiply
a * b
# division
a / b
# power
a ** b
# mod
a % b

tensor([[0.4411, 0.7565, 0.2239],
        [0.3023, 0.1784, 0.3346]])

In [ ]:
c = torch.tensor([1, -2, 3, -4])

In [ ]:
# abs
torch.abs(c)

tensor([1, 2, 3, 4])

In [ ]:
# negative
torch.neg(c)

tensor([-1,  2, -3,  4])

In [ ]:
d = torch.tensor([1.9, 2.3, 3.7, 4.4])

In [ ]:
# round
torch.round(d)

tensor([2., 2., 4., 4.])

In [ ]:
# ceil
torch.ceil(d)

tensor([2., 3., 4., 5.])

In [ ]:
# floor
torch.floor(d)

tensor([1., 2., 3., 4.])

In [ ]:
# clamp
torch.clamp(d, min=2, max=3)

tensor([2.0000, 2.3000, 3.0000, 3.0000])

### 3. Reduction operation

In [ ]:
e = torch.randint(size=(2,3), low=0, high=10, dtype=torch.float32)
e

tensor([[8., 0., 7.],
        [0., 0., 9.]])

### Explanation of `torch.randint(size=(2, 3), low=0, high=10, dtype=torch.float32)`

- `torch.randint(size=(2, 3), low=0, high=10, dtype=torch.float32)` creates a **2x3 tensor** with **random integers** drawn from the range `[0, 10)`, i.e., from 0 to 9.
- The `size=(2, 3)` argument defines the shape of the tensor with 2 rows and 3 columns.
- The `dtype=torch.float32` parameter explicitly sets the data type to `float32`, so the generated integers are cast as floating-point numbers.



In [ ]:
# sum
torch.sum(e)
# sum along columns
torch.sum(e, dim=0)
# sum along rows
torch.sum(e, dim=1)

tensor([15.,  9.])

In [ ]:
# mean
torch.mean(e)
# mean along col
torch.mean(e, dim=0)

tensor([4., 0., 8.])

In [ ]:
# median
torch.median(e)

tensor(0.)

In [ ]:
# max and min
torch.max(e)
torch.min(e)

tensor(0.)

In [ ]:
# product
torch.prod(e)

tensor(0.)

In [ ]:
# standard deviation
torch.std(e)

tensor(4.4272)

In [ ]:
# variance
torch.var(e)

tensor(19.6000)

In [ ]:
# argmax
torch.argmax(e)

tensor(5)

In [ ]:
# argmin
torch.argmin(e)

tensor(1)

### 4. Matrix operations

In [ ]:
f = torch.randint(size=(2,3), low=0, high=10)
g = torch.randint(size=(3,2), low=0, high=10)

print(f)
print(g)

tensor([[5, 7, 3],
        [9, 4, 0]])
tensor([[5, 7],
        [5, 9],
        [9, 7]])


In [ ]:
# matrix multiplcation
torch.matmul(f, g)

tensor([[ 87, 119],
        [ 65,  99]])

In [ ]:
vector1 = torch.tensor([1, 2])
vector2 = torch.tensor([3, 4])

# dot product
torch.dot(vector1, vector2)

tensor(11)

In [ ]:
# transpose
torch.transpose(f, 0, 1)

tensor([[5, 9],
        [7, 4],
        [3, 0]])

### Explanation of `torch.transpose(f, 0, 1)`

- `torch.transpose(f, 0, 1)` **swaps the specified dimensions** of the tensor `f`.
- Here, `0` and `1` represent the **row** and **column** dimensions, respectively.
- The operation changes the shape of the tensor by transposing its rows and columns.



In [ ]:
h = torch.randint(size=(3,3), low=0, high=10, dtype=torch.float32)
h

tensor([[5., 9., 8.],
        [9., 7., 9.],
        [2., 6., 7.]])

### Explanation of `torch.randint(size=(3, 3), low=0, high=10, dtype=torch.float32)`

- `torch.randint(size=(3, 3), low=0, high=10, dtype=torch.float32)` creates a **3x3 tensor** with **random integers** drawn from the range `[0, 10)`, i.e., from 0 to 9.
- The `size=(3, 3)` argument defines the shape of the tensor with 3 rows and 3 columns.
- The `dtype=torch.float32` parameter explicitly sets the data type to `float32`, so the generated integers are represented as floating-point numbers.



In [ ]:
# determinant
torch.det(h)

tensor(-110.)

In [ ]:
# inverse
torch.inverse(h)

tensor([[ 0.0455,  0.1364, -0.2273],
        [ 0.4091, -0.1727, -0.2455],
        [-0.3636,  0.1091,  0.4182]])

### 5. Comparison operations

In [ ]:
i = torch.randint(size=(2,3), low=0, high=10)
j = torch.randint(size=(2,3), low=0, high=10)

print(i)
print(j)

tensor([[7, 8, 3],
        [6, 1, 5]])
tensor([[5, 0, 4],
        [3, 8, 8]])


In [ ]:
# greater than
i > j
# less than
i < j
# equal to
i == j
# not equal to
i != j
# greater than equal to

# less than equal to

tensor([[True, True, True],
        [True, True, True]])

### 6. Special functions

In [ ]:
k = torch.randint(size=(2,3), low=0, high=10, dtype=torch.float32)
k

tensor([[3., 3., 5.],
        [0., 6., 4.]])

In [ ]:
# log
torch.log(k)

tensor([[1.0986, 1.0986, 1.6094],
        [  -inf, 1.7918, 1.3863]])

In [ ]:
# exp
torch.exp(k)

tensor([[ 20.0855,  20.0855, 148.4132],
        [  1.0000, 403.4288,  54.5981]])

In [ ]:
# sqrt
torch.sqrt(k)

tensor([[1.7321, 1.7321, 2.2361],
        [0.0000, 2.4495, 2.0000]])

In [ ]:
# sigmoid
torch.sigmoid(k)

tensor([[0.9526, 0.9526, 0.9933],
        [0.5000, 0.9975, 0.9820]])

In [ ]:
# softmax
torch.softmax(k, dim=0)

tensor([[0.9526, 0.0474, 0.7311],
        [0.0474, 0.9526, 0.2689]])

In [ ]:
# relu
torch.relu(k)

tensor([[3., 3., 5.],
        [0., 6., 4.]])

## Inplace Operations

In [ ]:
m = torch.rand(2,3)
n = torch.rand(2,3)

print(m)
print(n)

tensor([[0.6574, 0.3451, 0.0453],
        [0.9798, 0.5548, 0.6868]])
tensor([[0.4920, 0.0748, 0.9605],
        [0.3271, 0.0103, 0.9516]])


In [ ]:
m.add_(n)

tensor([[1.1494, 0.4199, 1.0058],
        [1.3069, 0.5650, 1.6384]])

### Explanation of `m.add_(n)`

- `m.add_(n)` performs **in-place addition** of tensor `n` to tensor `m`.
- The operation updates the values of `m` directly without creating a new tensor.
- The `_` at the end of `add_` indicates that it is an **in-place operation**, modifying the original tensor.




In [ ]:
m

tensor([[1.1494, 0.4199, 1.0058],
        [1.3069, 0.5650, 1.6384]])

In [ ]:
n

tensor([[0.4920, 0.0748, 0.9605],
        [0.3271, 0.0103, 0.9516]])

In [ ]:
torch.relu(m)

tensor([[1.1494, 0.4199, 1.0058],
        [1.3069, 0.5650, 1.6384]])

In [ ]:
m.relu_()

tensor([[1.1494, 0.4199, 1.0058],
        [1.3069, 0.5650, 1.6384]])

In [ ]:
m

tensor([[1.1494, 0.4199, 1.0058],
        [1.3069, 0.5650, 1.6384]])

## Copying a Tensor

In [ ]:
a = torch.rand(2,3)
a

tensor([[0.2855, 0.2324, 0.9141],
        [0.7668, 0.1659, 0.4393]])

In [ ]:
b = a

In [ ]:
b

tensor([[0.2855, 0.2324, 0.9141],
        [0.7668, 0.1659, 0.4393]])

In [ ]:
a[0][0] = 0

In [ ]:
a

tensor([[0.0000, 0.2324, 0.9141],
        [0.7668, 0.1659, 0.4393]])

In [ ]:
b

tensor([[0.0000, 0.2324, 0.9141],
        [0.7668, 0.1659, 0.4393]])

In [ ]:
id(a)

137647135422416

In [ ]:
id(b)

137647135422416

In [ ]:
b = a.clone()

### Explanation of `a.clone()`

- `a.clone()` creates a **deep copy** of the tensor `a`.
- The new tensor `b` will have the **same data and size** as `a`, but it is **stored at a different memory location**.
- This means changes to `b` will **not affect** `a`, and vice versa.



In [ ]:
a

tensor([[0.0000, 0.2324, 0.9141],
        [0.7668, 0.1659, 0.4393]])

In [ ]:
b

tensor([[0.0000, 0.2324, 0.9141],
        [0.7668, 0.1659, 0.4393]])

In [ ]:
a[0][0] = 10

In [ ]:
a

tensor([[10.0000,  0.2324,  0.9141],
        [ 0.7668,  0.1659,  0.4393]])

In [ ]:
b

tensor([[0.0000, 0.2324, 0.9141],
        [0.7668, 0.1659, 0.4393]])

In [ ]:
id(a)

137647135422416

In [ ]:
id(b)

137647135424432

# Performing Tensor Operations on GPU with PyTorch

In [ ]:
import torch

# Check if CUDA (GPU) is available
print(torch.cuda.is_available())  # Should return True if GPU is accessible


True


In [ ]:
# Define a variable to hold the GPU device
device = torch.device("cuda")


In [ ]:
# Create a tensor directly on the GPU
tensor_gpu = torch.rand(2, 3, device=device)
print(tensor_gpu)


tensor([[0.3563, 0.0303, 0.7088],
        [0.2009, 0.0224, 0.9896]], device='cuda:0')


In [ ]:
# Create a tensor on CPU
tensor_cpu = torch.tensor([[1, 2], [3, 4]])

# Move it to GPU
tensor_gpu = tensor_cpu.to(device)
print(tensor_gpu)


tensor([[1, 2],
        [3, 4]], device='cuda:0')


# 🆚 CPU vs GPU Speed Test (Matrix Multiplication)


In [ ]:
import time

# Create large CPU tensors
a_cpu = torch.rand(10000, 10000)
b_cpu = torch.rand(10000, 10000)

# Measure CPU time
start = time.time()
result_cpu = torch.matmul(a_cpu, b_cpu)
end = time.time()
print("CPU Time:", end - start, "seconds")


CPU Time: 17.674631357192993 seconds


In [ ]:
# Move tensors to GPU
a_gpu = a_cpu.to(device)
b_gpu = b_cpu.to(device)

# Warm-up GPU (helps in fair timing)
_ = torch.matmul(a_gpu, b_gpu)

# Measure GPU time
torch.cuda.synchronize()  # Wait for GPU to finish previous ops
start = time.time()
result_gpu = torch.matmul(a_gpu, b_gpu)
torch.cuda.synchronize()  # Wait again to get accurate time
end = time.time()
print("GPU Time:", end - start, "seconds")


GPU Time: 0.4775724411010742 seconds


# 🔁 Reshaping Tensors in PyTorch


## 📌 1. .reshape()
> This is the most direct way to reshape a tensor. Just like NumPy!

In [ ]:
import torch

a = torch.rand(4, 4)  # Shape: (4, 4)
b = a.reshape(2, 8)   # Reshape to (2, 8)

print(b.shape)  # Output: torch.Size([2, 8])


torch.Size([2, 8])


## 📌 2. .flatten()
> Use this to convert a multi-dimensional tensor into a 1D tensor.


In [ ]:
a = torch.rand(4, 4)
b = a.flatten()

print(b.shape)  # Output: torch.Size([16])


torch.Size([16])


## 📌 3. .permute()
> Use permute() when you want to reorder the dimensions of a tensor.

In [ ]:
b = torch.rand(2, 3, 4)  # Shape: (2, 3, 4)
c = b.permute(2, 0, 1)   # New shape: (4, 2, 3)

print(c.shape)  # Output: torch.Size([4, 2, 3])


torch.Size([4, 2, 3])


## 📌 4. .unsqueeze()
> Use unsqueeze() to add a new dimension at a specific position.

- Imagine a single image with shape (226, 226, 3). To feed it into a deep learning model that expects a batch, you must convert it into shape (1, 226, 226, 3):


In [ ]:
img = torch.rand(226, 226, 3)

# Add a new batch dimension at the front
img_batch = img.unsqueeze(0)

print(img_batch.shape)  # Output: torch.Size([1, 226, 226, 3])


torch.Size([1, 226, 226, 3])


## 📌 5. .squeeze()
> The inverse of unsqueeze() — it removes dimensions of size 1.

In [ ]:
x = torch.rand(1, 20)

# Remove the first dimension (size 1)
x_squeezed = x.squeeze(0)

print(x_squeezed.shape)  # Output: torch.Size([20])


torch.Size([20])


# 🔁 Converting Between NumPy Arrays and PyTorch Tensors
In real-world scenarios, you may need to switch between NumPy arrays and PyTorch tensors — especially when integrating existing NumPy-based workflows with PyTorch code (or vice versa).

Luckily, both conversions are easy and efficient.

## 📌 1. PyTorch Tensor ➡️ NumPy Array
You can convert a PyTorch tensor into a NumPy array using the .numpy() method:

In [ ]:
import torch

# Create a PyTorch tensor
a = torch.tensor([1, 2, 3])

# Convert it to a NumPy array
b = a.numpy()

print(type(b))  # Output: <class 'numpy.ndarray'>


<class 'numpy.ndarray'>


## 📌 2. NumPy Array ➡️ PyTorch Tensor
> You can convert a NumPy array to a PyTorch tensor using torch.from_numpy():



In [ ]:
import numpy as np

# Create a NumPy array
c = np.array([4, 5, 6])

# Convert it to a PyTorch tensor
d = torch.from_numpy(c)

print(type(d))  # Output: <class 'torch.Tensor'>


<class 'torch.Tensor'>
